### ISSUE TITLE
Order of arrayAll condition in LEFT JOIN to temporary table effects results

### ISSUE THREAD

**Issue by bkuschel** [ClickHouse/ClickHouse#71479]
Problem found in any CH **version > 24.3** where condition order determines the result when `LEFT JOIN`ing to a temporary table.

**How to reproduce**

* Which ClickHouse server version to use
Any version > 24.3

## `CREATE TABLE` statements for all tables involved
```sql
CREATE TABLE default.arrays_test
(
    s String,
    keys Array(String),
    values Array(String)
) 
ENGINE =MergeTree()
ORDER BY (s)
SETTINGS index_granularity = 8192;

CREATE TEMPORARY TABLE temp_table_f161e017
(
  a String
);
```

## Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary

```sql
INSERT INTO temp_table_f161e017
VALUES ('Hello'), ('World');

INSERT INTO default.arrays_test
VALUES ('Hello', ['abc1_app','abc1_appA'],['abc1_web', 'abc1_webA']), ('World', ['abc2_app','abc2_appA'],['abc2_web', 'abc2_webA']);
```

## Queries to run that lead to an unexpected result

[Query 1](https://fiddle.clickhouse.com/2f4ef8f4-9241-417a-90e4-6faaf73281ab):
```sql
SELECT s,a
FROM default.arrays_test LEFT JOIN temp_table_f161e017 on  default.arrays_test.s = a
WHERE (arrayExists((v) -> v  LIKE  '%abc%', keys) = 1 AND arrayAll((v) -> v  NOT LIKE '%abc%', values) = 1)
```
## Expected behavior
[Query 2 with the `WHERE` order gives correct result](https://fiddle.clickhouse.com/05d6c4c8-f88a-43ad-bfc8-d8b1a9782658):
```sql
SELECT s,a
FROM default.arrays_test LEFT JOIN temp_table_f161e017 on  default.arrays_test.s = a
WHERE (arrayAll((v) -> v  NOT LIKE '%abc%', values) = 1 AND arrayExists((v) -> v  LIKE  '%abc%', keys) = 1)
```

## A clear and concise description of what you expected to happen.

Query 1 should return the same result as Query 2

**Comment by Algunenano** [comment]
Seems like a bad pushdown in the join, since the filtered left table should be empty.

I've tried disabling join pushdown the hard way (removing the contents of splitActionsForJOINFilterPushDown) and it worked fine, so it's pretty much confirmed to be an issue there.

**Comment by SaltTan** [comment]
Seems fixed in master https://fiddle.clickhouse.com/d052c951-3b29-427e-99c0-4d05ef0f9157

**Comment by Algunenano** [comment]
Seems it was fixed by https://github.com/ClickHouse/ClickHouse/pull/58400 but only for the analyzer, so probably the pushdown is still broken.

**Comment by rschu1ze** [comment]
The non-analyzer path is pretty much dead by now. Let's close.

**Comment by larryluogit** [comment]
@rschu1ze @alexey-milovidov 
This issue is a regression in 24.8. Could this fix be backported to 24.8 LTS, please?

**Comment by rschu1ze** [comment]
@larryluogit #58400 is marked as "improvement", it may change the behavior in unexpected ways and this seems too risky to backport (especially into a LTS version).

**Comment by larryluogit** [comment]
How should we address the regression? Queries using the function are now returning incorrect results.

**Comment by rschu1ze** [comment]
Can confirm that #58400 resolved this, starting from v24.11, but it was more coincidence than intended.

Since the PR is about constant lambda arguments, I experimented with the `materialize` function (which turns it arguments non-const) but it did not make a difference for correctness (expectedly, since the arguments are already non-const). Replacing `arrayExists` and `arrayAll` by `arrayFilter` also didn't help.

What did actually help (and it is more of a random observation) was replacing `WHERE` by `PREWHERE` for this specific query. But it will not be a general workaround.
